File: ai_image_viewer/src/AI-Image-Viewer/collab/python/split_work.py
#!/usr/bin/env python3
"""
AI Image Viewer - Work Splitter Tool

This tool splits a source image folder into multiple work packages for collaborative analysis.
Each package contains:
- All 3 AI Image Viewer HTML files (MobileNet, EfficientNet, MediaPipe)
- A subset of images for analysis
- Ready-to-share bundle for team collaboration

Author: AI Image Viewer Team
"""

import os
import shutil
import math
import click
from pathlib import Path
from typing import List, Tuple
import json
from datetime import datetime


class WorkSplitter:
    """Handles splitting image datasets into collaborative work packages."""
    
    SUPPORTED_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp'}
    HTML_FILES = [
        'ai_image_viewer.html',
        'ai_image_viewer_efficientnet.html', 
        'ai_image_viewer_mediapipe.html'
    ]
    
    def __init__(self, source_folder: Path, output_folder: Path, num_splits: int):
        self.source_folder = Path(source_folder)
        self.output_folder = Path(output_folder)
        self.num_splits = num_splits
        self.project_root = self.find_project_root()
        
    def find_project_root(self) -> Path:
        """Find the project root containing HTML files."""
        current = Path(__file__).parent
        while current.parent != current:
            if all((current / html).exists() for html in self.HTML_FILES):
                return current
            current = current.parent
        raise FileNotFoundError("Could not find project root with HTML files")
    
    def get_image_files(self) -> List[Path]:
        """Get all supported image files from source folder."""
        image_files = []
        for ext in self.SUPPORTED_EXTENSIONS:
            image_files.extend(self.source_folder.glob(f"*{ext}"))
            image_files.extend(self.source_folder.glob(f"*{ext.upper()}"))
        return sorted(image_files)
    
    def split_images(self, image_files: List[Path]) -> List[List[Path]]:
        """Split image files into roughly equal chunks."""
        if not image_files:
            raise ValueError("No image files found in source folder")
            
        chunk_size = math.ceil(len(image_files) / self.num_splits)
        chunks = []
        
        for i in range(self.num_splits):
            start_idx = i * chunk_size
            end_idx = min((i + 1) * chunk_size, len(image_files))
            chunk = image_files[start_idx:end_idx]
            if chunk:  # Only add non-empty chunks
                chunks.append(chunk)
                
        return chunks
    
    def create_work_package(self, chunk_id: int, image_chunk: List[Path]) -> Path:
        """Create a complete work package with HTML files and images."""
        package_name = f"work-package-{chunk_id:03d}"
        package_path = self.output_folder / package_name
        
        # Create package directory
        package_path.mkdir(parents=True, exist_ok=True)
        
        # Copy HTML files
        for html_file in self.HTML_FILES:
            src_path = self.project_root / html_file
            dst_path = package_path / html_file
            shutil.copy2(src_path, dst_path)
            click.echo(f"  ✓ Copied {html_file}")
        
        # Copy docs folder if it exists
        docs_src = self.project_root / "docs"
        if docs_src.exists():
            docs_dst = package_path / "docs"
            shutil.copytree(docs_src, docs_dst, dirs_exist_ok=True)
            click.echo(f"  ✓ Copied docs folder")
        
        # Create images subfolder and copy images
        images_path = package_path / "images"
        images_path.mkdir(exist_ok=True)
        
        for img_file in image_chunk:
            dst_path = images_path / img_file.name
            shutil.copy2(img_file, dst_path)
            
        click.echo(f"  ✓ Copied {len(image_chunk)} images")
        
        # Create package manifest
        self.create_manifest(package_path, chunk_id, image_chunk)
        
        return package_path
    
    def create_manifest(self, package_path: Path, chunk_id: int, image_chunk: List[Path]):
        """Create a manifest file describing the package contents."""
        manifest = {
            "package_info": {
                "id": chunk_id,
                "name": f"work-package-{chunk_id:03d}",
                "created": datetime.now().isoformat(),
                "total_packages": self.num_splits,
                "image_count": len(image_chunk)
            },
            "html_files": self.HTML_FILES,
            "images": [img.name for img in image_chunk],
            "instructions": {
                "1": "Open any of the HTML files in your browser",
                "2": "Load images from the 'images' folder",
                "3": "Click 'AI Analyze' to process your assigned images",
                "4": "Export metadata when analysis is complete",
                "5": "Share the exported JSON file back to the project lead"
            }
        }
        
        manifest_path = package_path / "package-manifest.json"
        with open(manifest_path, 'w', encoding='utf-8') as f:
            json.dump(manifest, f, indent=2, ensure_ascii=False)
        
        click.echo(f"  ✓ Created package manifest")
    
    def split(self) -> List[Path]:
        """Main method to split the work."""
        click.echo(f"🔍 Analyzing source folder: {self.source_folder}")
        
        # Get all image files
        image_files = self.get_image_files()
        if not image_files:
            raise click.ClickException(f"No image files found in {self.source_folder}")
        
        click.echo(f"📸 Found {len(image_files)} image files")
        
        # Split into chunks
        chunks = self.split_images(image_files)
        actual_splits = len(chunks)
        
        if actual_splits != self.num_splits:
            click.echo(f"⚠️  Created {actual_splits} packages instead of {self.num_splits} (insufficient images)")
        
        click.echo(f"📦 Creating {actual_splits} work packages in: {self.output_folder}")
        
        # Create output directory
        self.output_folder.mkdir(parents=True, exist_ok=True)
        
        # Create each work package
        created_packages = []
        for i, chunk in enumerate(chunks):
            click.echo(f"\n📦 Creating package {i+1}/{actual_splits}:")
            package_path = self.create_work_package(i+1, chunk)
            created_packages.append(package_path)
            click.echo(f"   Package: {package_path.name} ({len(chunk)} images)")
        
        return created_packages


@click.command()
@click.option('--source', '-s', 
              type=click.Path(exists=True, file_okay=False, dir_okay=True, path_type=Path),
              required=True,
              help='Source folder containing images to split')
@click.option('--num-splits', '-n',
              type=int,
              required=True,
              help='Number of work packages to create')
@click.option('--output', '-o',
              type=click.Path(path_type=Path),
              default=None,
              help='Output folder for work packages (default: ./split/)')
@click.version_option(version='1.0.0', prog_name='AI Image Viewer Work Splitter')
def main(source: Path, num_splits: int, output: Path):
    """
    🧠 AI Image Viewer - Work Splitter Tool
    
    Split a source image folder into multiple work packages for team collaboration.
    Each package contains HTML viewers and a subset of images for analysis.
    
    Examples:
    
    \b
    # Split 1000 images into 5 packages for 5 team members
    python split_work.py -s ./big-dataset -n 5
    
    \b
    # Split with custom output location
    python split_work.py -s /path/to/images -n 3 -o /path/to/packages
    """
    
    if num_splits < 1:
        raise click.BadParameter("Number of splits must be at least 1")
    
    if num_splits > 100:
        raise click.BadParameter("Number of splits cannot exceed 100 (too many packages)")
    
    # Set default output location
    if output is None:
        output = Path(__file__).parent / "split"
    
    try:
        click.echo("🧠 AI Image Viewer - Work Splitter Tool")
        click.echo("=" * 45)
        
        # Initialize splitter
        splitter = WorkSplitter(source, output, num_splits)
        
        # Perform the split
        packages = splitter.split()
        
        click.echo(f"\n✅ Successfully created {len(packages)} work packages!")
        click.echo(f"📁 Output location: {output}")
        
        # Show summary
        click.echo(f"\n📋 Package Summary:")
        for i, package in enumerate(packages):
            manifest_path = package / "package-manifest.json"
            if manifest_path.exists():
                with open(manifest_path, 'r') as f:
                    manifest = json.load(f)
                image_count = manifest['package_info']['image_count']
                click.echo(f"   {package.name}: {image_count} images")
        
        click.echo(f"\n🚀 Next Steps:")
        click.echo(f"1. Compress each package: tar -czf package.tar.gz work-package-001/")
        click.echo(f"2. Share packages with team members")
        click.echo(f"3. Team members analyze their assigned images")
        click.echo(f"4. Use join_work.py to merge results when complete")
        
    except Exception as e:
        click.echo(f"❌ Error: {e}", err=True)
        raise click.Abort()


if __name__ == '__main__':
    main()

File: ai_image_viewer/src/AI-Image-Viewer/collab/js/split-work.js
#!/usr/bin/env node

/**
 * AI Image Viewer - Work Splitter Tool (JavaScript)
 * 
 * This tool splits a source image folder into multiple work packages for collaborative analysis.
 * Each package contains:
 * - All 3 AI Image Viewer HTML files (MobileNet, EfficientNet, MediaPipe)
 * - A subset of images for analysis
 * - Ready-to-share bundle for team collaboration
 * 
 * Author: AI Image Viewer Team
 */

import { Command } from 'commander';
import fs from 'fs-extra';
import path from 'path';
import { glob } from 'glob';
import chalk from 'chalk';

class WorkSplitter {
    constructor(sourceFolder, outputFolder, numSplits) {
        this.sourceFolder = path.resolve(sourceFolder);
        this.outputFolder = path.resolve(outputFolder);
        this.numSplits = numSplits;
        this.supportedExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp'];
        this.htmlFiles = [
            'ai_image_viewer.html',
            'ai_image_viewer_efficientnet.html',
            'ai_image_viewer_mediapipe.html'
        ];
        this.projectRoot = this.findProjectRoot();
    }

    findProjectRoot() {
        let current = path.dirname(new URL(import.meta.url).pathname);
        while (current !== path.dirname(current)) {
            const hasAllHtml = this.htmlFiles.every(htmlFile => 
                fs.existsSync(path.join(current, htmlFile))
            );
            if (hasAllHtml) {
                return current;
            }
            current = path.dirname(current);
        }
        throw new Error('Could not find project root with HTML files');
    }

    async getImageFiles() {
        const patterns = this.supportedExtensions.flatMap(ext => [
            path.join(this.sourceFolder, `*${ext}`),
            path.join(this.sourceFolder, `*${ext.toUpperCase()}`)
        ]);
        
        const imageFiles = [];
        for (const pattern of patterns) {
            const matches = await glob(pattern);
            imageFiles.push(...matches);
        }
        
        return [...new Set(imageFiles)].sort(); // Remove duplicates and sort
    }

    splitImages(imageFiles) {
        if (imageFiles.length === 0) {
            throw new Error('No image files found in source folder');
        }

        const chunkSize = Math.ceil(imageFiles.length / this.numSplits);
        const chunks = [];

        for (let i = 0; i < this.numSplits; i++) {
            const startIdx = i * chunkSize;
            const endIdx = Math.min((i + 1) * chunkSize, imageFiles.length);
            const chunk = imageFiles.slice(startIdx, endIdx);
            if (chunk.length > 0) {
                chunks.push(chunk);
            }
        }

        return chunks;
    }

    async createWorkPackage(chunkId, imageChunk) {
        const packageName = `work-package-${chunkId.toString().padStart(3, '0')}`;
        const packagePath = path.join(this.outputFolder, packageName);

        // Create package directory
        await fs.ensureDir(packagePath);

        // Copy HTML files
        for (const htmlFile of this.htmlFiles) {
            const srcPath = path.join(this.projectRoot, htmlFile);
            const dstPath = path.join(packagePath, htmlFile);
            if (await fs.pathExists(srcPath)) {
                await fs.copy(srcPath, dstPath);
                console.log(chalk.green(`  ✓ Copied ${htmlFile}`));
            }
        }

        // Copy docs folder if it exists
        const docsSrc = path.join(this.projectRoot, 'docs');
        if (await fs.pathExists(docsSrc)) {
            const docsDst = path.join(packagePath, 'docs');
            await fs.copy(docsSrc, docsDst);
            console.log(chalk.green('  ✓ Copied docs folder'));
        }

        // Create images subfolder and copy images
        const imagesPath = path.join(packagePath, 'images');
        await fs.ensureDir(imagesPath);

        for (const imgFile of imageChunk) {
            const fileName = path.basename(imgFile);
            const dstPath = path.join(imagesPath, fileName);
            await fs.copy(imgFile, dstPath);
        }

        console.log(chalk.green(`  ✓ Copied ${imageChunk.length} images`));

        // Create package manifest
        await this.createManifest(packagePath, chunkId, imageChunk);

        return packagePath;
    }

    async createManifest(packagePath, chunkId, imageChunk) {
        const manifest = {
            package_info: {
                id: chunkId,
                name: `work-package-${chunkId.toString().padStart(3, '0')}`,
                created: new Date().toISOString(),
                total_packages: this.numSplits,
                image_count: imageChunk.length
            },
            html_files: this.htmlFiles,
            images: imageChunk.map(img => path.basename(img)),
            instructions: {
                "1": "Open any of the HTML files in your browser",
                "2": "Load images from the 'images' folder",
                "3": "Click 'AI Analyze' to process your assigned images",
                "4": "Export metadata when analysis is complete",
                "5": "Share the exported JSON file back to the project lead"
            }
        };

        const manifestPath = path.join(packagePath, 'package-manifest.json');
        await fs.writeJSON(manifestPath, manifest, { spaces: 2 });
        console.log(chalk.green('  ✓ Created package manifest'));
    }

    async split() {
        console.log(chalk.cyan(`🔍 Analyzing source folder: ${this.sourceFolder}`));

        // Get all image files
        const imageFiles = await this.getImageFiles();
        if (imageFiles.length === 0) {
            throw new Error(`No image files found in ${this.sourceFolder}`);
        }

        console.log(chalk.blue(`📸 Found ${imageFiles.length} image files`));

        // Split into chunks
        const chunks = this.splitImages(imageFiles);
        const actualSplits = chunks.length;

        if (actualSplits !== this.numSplits) {
            console.log(chalk.yellow(`⚠️  Created ${actualSplits} packages instead of ${this.numSplits} (insufficient images)`));
        }

        console.log(chalk.blue(`📦 Creating ${actualSplits} work packages in: ${this.outputFolder}`));

        // Create output directory
        await fs.ensureDir(this.outputFolder);

        // Create each work package
        const createdPackages = [];
        for (let i = 0; i < chunks.length; i++) {
            const chunk = chunks[i];
            console.log(chalk.blue(`\n📦 Creating package ${i + 1}/${actualSplits}:`));
            const packagePath = await this.createWorkPackage(i + 1, chunk);
            createdPackages.push(packagePath);
            console.log(chalk.green(`   Package: ${path.basename(packagePath)} (${chunk.length} images)`));
        }

        return createdPackages;
    }
}

// CLI Setup
const program = new Command();

program
    .name('split-work')
    .description('🧠 AI Image Viewer - Work Splitter Tool\n\nSplit a source image folder into multiple work packages for team collaboration.')
    .version('1.0.0')
    .requiredOption('-s, --source <directory>', 'Source folder containing images to split')
    .requiredOption('-n, --num-splits <number>', 'Number of work packages to create', parseInt)
    .option('-o, --output <directory>', 'Output folder for work packages (default: ./split/)')
    .addHelpText('after', `
Examples:
  $ node split-work.js -s ./big-dataset -n 5
  $ node split-work.js -s /path/to/images -n 3 -o /path/to/packages
  $ npx split-work -s ./dataset -n 4
`);

program.parse();

const options = program.opts();

// Validation
if (options.numSplits < 1) {
    console.error(chalk.red('❌ Number of splits must be at least 1'));
    process.exit(1);
}

if (options.numSplits > 100) {
    console.error(chalk.red('❌ Number of splits cannot exceed 100 (too many packages)'));
    process.exit(1);
}

if (!await fs.pathExists(options.source)) {
    console.error(chalk.red(`❌ Source folder does not exist: ${options.source}`));
    process.exit(1);
}

// Set default output location
const outputFolder = options.output || path.join(process.cwd(), 'split');

// Main execution
async function main() {
    try {
        console.log(chalk.bold.blue('🧠 AI Image Viewer - Work Splitter Tool'));
        console.log(chalk.blue('=' * 45));

        // Initialize splitter
        const splitter = new WorkSplitter(options.source, outputFolder, options.numSplits);

        // Perform the split
        const packages = await splitter.split();

        console.log(chalk.green(`\n✅ Successfully created ${packages.length} work packages!`));
        console.log(chalk.blue(`📁 Output location: ${outputFolder}`));

        // Show summary
        console.log(chalk.blue('\n📋 Package Summary:'));
        for (const packagePath of packages) {
            const manifestPath = path.join(packagePath, 'package-manifest.json');
            if (await fs.pathExists(manifestPath)) {
                const manifest = await fs.readJSON(manifestPath);
                const imageCount = manifest.package_info.image_count;
                const packageName = path.basename(packagePath);
                console.log(chalk.white(`   ${packageName}: ${imageCount} images`));
            }
        }

        console.log(chalk.green('\n🚀 Next Steps:'));
        console.log(chalk.white('1. Compress each package: tar -czf package.tar.gz work-package-001/'));
        console.log(chalk.white('2. Share packages with team members'));
        console.log(chalk.white('3. Team members analyze their assigned images'));
        console.log(chalk.white('4. Use join-work to merge results when complete'));

    } catch (error) {
        console.error(chalk.red(`❌ Error: ${error.message}`));
        process.exit(1);
    }
}

main();